autogen/agentchat/contrib/postgresql_agent.py

```diff
@@ -0,0 +1,153 @@
+
+class PostgreSqlQueryExecutor(CodeExecutor):
+    def __init__(self, dsn: str, **kwargs):
+        """(Experimental) A code executor class that executes SQL SELECT query statement.
+
+        **This will execute SQL SELECT query statement on PostgreSQL server.**
+
+        The code blocks are executed or save in the order they are received.
+        SQL statements is sanitized against `DROP`, `UPDATE`, `DELETE`, and `INSERT` statement to prevent updates to database.
+        This may not be enough, to add more prevention, please set the database user permission on your database server accordingly.
+
+        Args:
+            dsn (str): connection string to specify the PostgreSQL server.
+        """
+        self.dsn = dsn
+
+    @property
+    def code_extractor(self) -> CodeExtractor:
+        return MarkdownCodeExtractor()
+
+    @staticmethod
+    def sanitize_command(code: str) -> None:
+        """
+        Sanitize the code block to prevent `DROP`, `UPDATE`, `DELETE`, and `INSERT` statement .
+        """
+        dangerous_patterns = [
+            (r"\bdrop\b", "Use of 'drop' query is not allowed."),
+            (r"\binsert\b", "Use of 'insert' query is not allowed."),
+            (r"\bupdate\b", "Use of 'update' query is not allowed."),
+            (r"\bdelete\b", "Use of 'delete' query is not allowed."),
+        ]
+        for pattern, message in dangerous_patterns:
+            if re.search(pattern, code, re.IGNORECASE):
+                raise ValueError(f"Potentially dangerous command detected: {message}")
+
+    def execute_code_blocks(self, code_blocks: List[CodeBlock]) -> CodeResult:
+        """(Experimental) Execute the code blocks and return the result.
+
+        Args:
+            code_blocks (List[CodeBlock]): The code blocks to execute.
+
+        Returns:
+            CodeResult: The result of the code execution."""
+        logs_all = ""
+        exitcode = -1
+        for idx, code_block in enumerate(code_blocks, start=1):
+            lang, code = code_block.language, code_block.code
+            lang = lang.lower()
+
+            if lang != "sql":
+                logs_all += "\n" + f"Skipping execution: language not specified (code block #{idx})."
+                continue
+
+            try:
+                PostgreSqlQueryExecutor.sanitize_command(code)
+
+                with pg.connect(self.dsn, row_factory=dict_row) as cnn:
+                    with cnn.cursor() as cr:
+                        cr.execute(code)
+                        cres = cr.fetchall()
+                        res_json = json.dumps(cres, default=str)
+            except Exception as e:
+                exitcode = -1
+                logs_all += f"\nError: {str(e)}"
+                break
+            exitcode = 0
+            logs_all += res_json
+        return CodeResult(exit_code=exitcode, output=logs_all)
+
+    def restart(self) -> None:
+        return
+
+class PostgreSqlAgent(ConversableAgent):
```

Comment:
If we go even further,  this class is not very much needed once we extend the existing code executor to support `sql` language. The `cnn` could be passed into the executor.